### PR TITLE
Refactor destroy functions and save workspaces when there's no outputs

### PR DESCRIPTION
**Pull request by RyanDwyer**
This changes the destroy functions to the following:

* `output_begin_destroy`
* `output_destroy`
* `workspace_begin_destroy`
* `workspace_destroy`
* `container_begin_destroy`
* `container_destroy`
* `view_begin_destroy`
* `view_destroy`

The terminology was `destroy` and `free`, and it has been changed to `begin_destroy` and `destroy` respectively.

When the last output is disconnected, its workspaces will now be stashed in the root. Upon connection of a new output they will be restored.

There is a new function `workspace_consider_destroy` which decides whether the given workspace should be destroyed or not (ie. empty and not visible).

Calling `container_begin_destroy` will no longer automatically reap the parents. In some places we want to reap the parents and in some we don't, so this is left to the caller.

`container_reap_empty_recursive` and `container_reap_empty` have been combined into one function and it will recurse up the tree.

This needs testing with output hotplugging, and unplugging the last output to see what happens.

Despite adding more functions, I've somehow managed to delete more code than I've added :)

### CODE REVIEW

File: sway/tree/container.c [hunk 1]
```diff
@@ -104,209 +104,52 @@ struct sway_container *container_create(enum sway_container_type type) {
...
-			while (output->children->length) {
-				struct sway_container *workspace = output->children->items[0];
-
-				struct sway_container *new_output =
-					workspace_output_get_highest_available(workspace, output);
-				if (!new_output) {
-					new_output = fallback_output;
-					workspace_output_add_priority(workspace, new_output);
-				}
+	free(con->name);
```
Comment (emersion) [inline sway/tree/container.c]: This is duplicated with:

https://github.com/swaywm/sway/pull/2499/files#diff-101a892d1de03398ff5a39989c2332dcR162

Maybe we should consider calling this function from the other?

Comment (RyanDwyer) [reply sway/tree/container.c]: The plan is for the output and workspace destroy functions to eventually destroy a `sway_output` and `sway_workspace` (and likewise for their create functions). So while there's some code duplication at the moment, it's only temporary.

The purpose behind all these refactor PRs is to move things into their proper locations so that I can demote `sway_container` without it being too big of a PR.

**Comment by RyanDwyer** [comment]
`begin_destroy` removes it from the pending tree and marks it as destroying. Then it gets included in the next transaction. Once that transaction has finished (ie. the thing is no longer rendered on screen), the `destroy` function is called to actually free all the resources for it. We had the same thing already, but `container_destroy` was the first function and `container_free` was the second function.

**Comment by SirCmpwn** [comment]
Got it.

**Comment by RyanDwyer** [comment]
I've been running this branch as my daily for the last 24 hours and haven't had any problems. I also tested hotplugging and didn't notice any issues, though I didn't hotplug down to zero outputs.

**Comment by RyanDwyer** [comment]
Bump.

**Comment by SirCmpwn** [comment]
Thanks!
